test/pynrn/test_nrntest_fast.py

```diff
@@ -0,0 +1,304 @@
+"""
+Tests that used to live in the fast/ subdirectory of the
+https://github.com/neuronsimulator/nrntest repository
+"""
+import math
+import numpy as np
+import os
+import pytest
+from neuron import h
+from neuron.tests.utils import (
+    cvode_enabled,
+    cvode_use_long_double,
+    hh_table_disabled,
+    num_threads,
+    parallel_context,
+)
+from neuron.tests.utils.checkresult import Chk
+
+h.load_file("stdrun.hoc")
+
+
+@pytest.fixture(scope="module")
+def chk():
+    """Manage access to JSON reference data."""
+    dir_path = os.path.dirname(os.path.realpath(__file__))
+    checker = Chk(os.path.join(dir_path, "test_nrntest_fast.json"))
+    yield checker
+    # Save results to disk if they've changed; this is called after all tests
+    # using chk have executed
+    checker.save()
+
+
+class Cell:
+    """Cell class used for test_t13 and augmented for test_t14."""
+
+    def __init__(self, id):
+        self.id = id
+        self.soma = h.Section(name="soma", cell=self)
+        self.soma.L = 10
+        self.soma.diam = 100 / (math.pi * self.soma.L)
+        self.soma.insert("hh")
+        self.s = h.IClamp(self.soma(0.5))
+        self.s.dur = 0.1 + 0.1 * id
+        self.s.amp = 0.2 + 0.1 * id
+        self.vv = h.Vector()
+        self.vv.record(self.soma(0.5)._ref_v)
+        self.tv = h.Vector()
+        self.tv.record(h._ref_t)
+
+    def data(self):
+        return {"t": list(self.tv), "v": list(self.vv)}
+
+    def pr(self, file=None):
+        print("\n{}".format(self), file=file)
+        assert self.tv.size() == self.vv.size()
+        for t, v in zip(self.tv, self.vv):
+            print("{:.15g} {:.15g}".format(t, v), file=file)
+
+    def __str__(self):
+        return "Cell[{:d}]".format(self.id)
+
+
+# Run t13 and t13 with 1 thread and with 3 threads if MPI is disabled. With
+# MPI enabled then multithreading is not supported, so just use 1 thread.
+pc = h.ParallelContext()
+this_rank, num_ranks = pc.id(), pc.nhost()
+thread_values = [1] if num_ranks > 1 else [1, 3]
+t13_methods = ["fixed"]
+if num_ranks == 1:
+    # There are diffs if we try and run t13 + cvode with multiple ranks.
+    # TODO: verify why/check with Michael, maybe because the different cells
+    # are only coupled via a global CVODE instance?
+    t13_methods += ["cvode", "cvode_long_double"]
```

Comment:
@nrnhines in the end I have made quite a few changes here since your previous review.

CodeCov encouraged me to try and cover some nrnvector_*.cpp codepaths, which led to trying to execute these tests (the old t13.hoc and t14.hoc models) with MPI too. For t14.hoc I fairly straightforwardly got ~the same results with MPI (2 ranks, 1 thread each), but for t13.hoc I had to disable the CVODE configurations with MPI.

I'm not quite clear on whether this expected to work or not in the sense of CVODE actually choosing the same series of time steps up to ~numerical precision. Can you clarify?

Reply:
> Can you clarify?

Probably not :)

For t13.hoc I did expect the cvode results with use_long_double and three threads to be the same over time (defined by t13.stdout within 1e-8) . I see someone has added the alternative t13.stdout.ci.nvhpc.21.9.

I see that t13.hoc does not compare 1 thread results with 3 thread results (which is the whole point of use_long_double!)
I wonder if I just gave up the attempt. There has always been a question in regard to the determinism of how cvode selects dt. It is possible (i've done it before) but painful to answer the question: When does cvode first differ and why? (i.e. where/which arithmetic operation causes a state or dt to differ.) I went through that for a difference of a model on windows and linux.

Reply:
One can force cvode to integrate silently using a specified series of time intervals. I would expect the ~numerical precision to be better. For less precision (because of piece wise linear) one can use vector interpolation to the same times as the standard.

Reply:
In my "new" versions of these tests I added a 1-thread vs. 3-thread comparison to t13.
(I only store a reference for 1-thread on disk, so the 1-vs-3 comparison is from the same binaries, while the 1-vs-reference comparison is comparing across compilers/architectures/...)

I did spend a bit of time looking at the sources of the diffs, but it seemed to me that cvode is good at amplifying diffs at the limit of `double` precision into much larger shifts in the chosen time values. I already implemented the interpolation you mentioned, which was enough to keep the diffs at the O(1e-7) relative level, which seemed good enough to me.

But: what I was trying to ask was about something that the original `t13.hoc` didn't do at all, namely distributing the same three cells across two MPI ranks (so two cells on one rank and one on the other) and comparing to the reference results (produced with all three cells on the same rank). The results are very different, but I'm not quite clear on whether this is because I made a mistake, or if it's expected (e.g. because the timesteps for all cells on a rank have to match, and multiple ranks changes what that restriction means, or something)?

Reply:
I was essentially just trying to extend the tests minimally to cover these lines:
https://github.com/neuronsimulator/nrn/blob/10e8f743ec95c79441fbee6704132e4420ef7bb4/src/nrncvode/cvodeobj.cpp#L813-L817

Reply:
Weere you successful in covering those lines (but at the expense of differing results)?

The intention for the global variable time step method was that results for different nhost and nthread be numerically identical except for numerical differences in the summation reduction due to different addition reduction orderings. (Note that this does not extend to multisplit which has much larger roundoff error accumulation for reasons having to do with very different gaussian elimination row orderings and even non-tree topology submatrices within a single tree). I was definitely hoping that use_long_double would effectively eliminate the reduction roundoff error. (and would be the user method of choice to determine if differences betweeen nhost and nthread were only a reduction problem instead of implicating something more substantive).  Perhaps it makes sense for me to employ that "possible but painful" method to see the cause of the first difference.

Reply:
in another train of thoughts, @nrnhines IIRC you were saying we should drop `N_VNew_NrnParallelLD` in favor of upstream SUNDIALS? Would that  alleviate some of the concerns?

Reply:
I believe I successfully covered them via my replacement for `t14.hoc`, and got ~identical results.
With `t13.hoc` I get wildly different results. From the above, it looks like this is unexpected and I should look into it further. (it should be easy to reproduce: on this branch, change the `if num_ranks == 1:` this comment is attached to to be `if True:`, then build and run `ctest -R parallel::nrntest -V`)

Reply:
This was fixed in 2e2badfb54634819231096d0ca5feae4377688d1 thanks to @nrnhines's investigation behind the scenes.